1-base/src/main/java/com/example/android/wearable/complicationsdataprovider/ComplicationTapBroadcastReceiver.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.complicationsdataprovider
+
+import android.app.PendingIntent
+import android.content.BroadcastReceiver
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import androidx.core.content.edit
+
+/**
+ * Simple [BroadcastReceiver] subclass for asynchronously incrementing an integer for any
+ * complication id triggered via TapAction on complication. Also, provides static method to create
+ * a [PendingIntent] that triggers this receiver.
+ */
+class ComplicationTapBroadcastReceiver : BroadcastReceiver() {
+    override fun onReceive(context: Context, intent: Intent) {
+        val extras = intent.extras ?: return
+        val provider = extras.getParcelable<ComponentName>(EXTRA_PROVIDER_COMPONENT) ?: return
+        val complicationId = extras.getInt(EXTRA_COMPLICATION_ID)
+
+        // Retrieve data via SharedPreferences.
+        val preferenceKey = getPreferenceKey(provider, complicationId)
+        val sharedPreferences =
+            context.getSharedPreferences(COMPLICATION_PROVIDER_PREFERENCES_FILE_KEY, 0)
+        var value = sharedPreferences.getInt(preferenceKey, 0)
+
+        // Update data for complication.
+        value = (value + 1) % MAX_NUMBER
+        sharedPreferences.edit { putInt(preferenceKey, value) }
```

Comment:
Don't you still have to do an `apply()` here?

Reply:
Not when you use android-ktx. See https://developer.android.com/kotlin/ktx